adding getter for an excel file (ONS)

cmbrennan002 opened the pull request:
---

# Description

This PR explores the salary ojo output, and creates tests to:

- Check for internal consistency
- Create a merged file of the fields needed to check OJO salary against ASHE, and save to S3
- Check against 2023 ASHE data

Fixes # 18 (salary EDA)

# Instructions for Reviewer

You can run the internal tests from your root folder by running:

python -m dap_job_quality.pipeline.salary_internal_tests>

The OJO file takes some time (40 mins) to create, so I'd suggest you just read through the code, but if you want to test it you can run: 

python -m dap_job_quality.utils.create_df_for_salary_analysis

The ASHE comparison is currently a notebook, to check it download the notebook and explore:

dap_job_quality.notebooks.Salary_comparison_ASHE.ipynb

Please pay special attention to ...

# Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [ ] I have requested a code review

cmbrennan002 committed:
adding getter for an excel file (ONS)

cmbrennan002 committed:
extra requirement for excel getter

cmbrennan002 committed:
adding internal salary tests

cmbrennan002 committed:
create df of relevant ojo data for salary analysis

cmbrennan002 committed:
adding notebook of ashe comparison for initial checks

cmbrennan002 committed:
adding notebook of ashe comparison for initial checks

cmbrennan002 committed:
Fixing minor errors in notebook

RFOxbury commented on dap_job_quality/pipeline/salary_internal_tests.py:
```diff
@@ -0,0 +1,53 @@
+            ojo_sal["max_annualised_salary"] - ojo_sal["min_annualised_salary"]
+        )
+        ojo_sal["sal_diff"].fillna(0, inplace=True)
+        self.assertTrue(
```
You could potentially make these tests more helpful by getting them to print out IDs where the test failed? Or even extend this to saving a list of failed IDs for each test.

```suggestion
        failed_ids = ojo_sal[ojo_sal["sal_diff"] < 0]["id"]
        self.assertTrue(
            failed_ids.empty,
            f"Minimum salary greater than maximum salary for IDs: {failed_ids.tolist()}",
        )
```

RFOxbury commented on dap_job_quality/pipeline/salary_internal_tests.py:
```diff
@@ -0,0 +1,53 @@
+        )
+
+    def test_max_diff(self):
+        self.assertTrue(
```
```suggestion
        condition = ojo_sal["max_annualised_salary"] > ojo_sal["min_annualised_salary"] * 10
        failed_ids = ojo_sal[condition]["id"]
        self.assertTrue(
            failed_ids.empty,
            f"Maximum salary more than 10 times the minimum salary for IDs: {failed_ids.tolist()}",
        )
```

RFOxbury commented on dap_job_quality/pipeline/salary_internal_tests.py:
```diff
@@ -0,0 +1,53 @@
+        )
+
+    def test_max_salary(self):
+        self.assertTrue(
```
```suggestion
        condition = ojo_sal["max_annualised_salary"] > 500000
        failed_ids = ojo_sal[condition]["id"]
        self.assertTrue(
            failed_ids.empty,
            f"Maximum salary greater than 500000 for IDs: {failed_ids.tolist()}",
        )
```

RFOxbury commented on dap_job_quality/utils/create_df_for_salary_analysis.py:
```diff
@@ -0,0 +1,103 @@
+
+
+# Current OJO data paths
+occupation_measures_s3 = "outputs/data/ojo_application/extracted_green_measures/20240213/ojo_large_sample_occupation_green_measures_production_true.csv"
```
This is such a tiny thing but I think as global constants these could/should be all caps

RFOxbury commented on requirements.txt:
```diff
@@ -7,3 +7,5 @@ sentence-transformers
 toolz
 plac
 srsly
```
Please could you add seaborn?

RFOxbury commented on dap_job_quality/utils/create_df_for_salary_analysis.py:
```diff
@@ -0,0 +1,103 @@
+    ojo_df["SOC_clean"] = ojo_df["SOC"].fillna("None")
+    ojo_df["SOC_clean"] = ojo_df["SOC_clean"].apply(lambda x: x.replace("nan", "None"))
+    ojo_df["SOC_clean"] = ojo_df["SOC_clean"].apply(
+        lambda x: eval(x) if isinstance(x, str) else x
```
This is nice! I didn't know about eval. I think I've scratched my head about how to get dictionaries and lists out of pandas df cells before!

RFOxbury commented on dap_job_quality/utils/create_df_for_salary_analysis.py:
```diff
@@ -0,0 +1,103 @@
+            f"File saved to {BUCKET_NAME}/job_quality/salary_analysis/salary_analysis_df.csv"
+        )
+
+    return ojo_df
```
This df has NAs for the salary in about 25% of rows. The file `outputs/data/ojo_application/deduplicated_sample/all_salaries_data_sample.parquet` is 3,382,859 rows while the cleaned salaries df has information for 4,653,782 unique job adverts. Did you know about this / does Jack know? Does this mean we're missing salary information for some job adverts (about 1 million)?

RFOxbury reviewed (approved):
Thanks Clare!

To review this PR, I:
- ran `python -m dap_job_quality.pipeline.salary_internal_tests`
- Looked through `create_df_for_salary_analysis.py` but did not run it
- Ran through the notebook
- Investigated the dataframe `'job_quality/salary_analysis/salary_analysis_df.csv'`

I have suggested some small edits but I don't think they're that important. You could maybe make a subfolder called `salary/` inside `pipeline/` to put the salary scripts in for now and to keep them all together, but up to you.

The only weird thing to flag is that salary information seems to be missing for some job adverts? I don't know if that's because the Prinz file `"outputs/data/ojo_application/deduplicated_sample/all_salaries_data_sample.parquet"` is not complete. Sorry if this has already come up somewhere and I missed it.

cmbrennan002 committed:
updated internal salary tests and file structure

cmbrennan002 committed:
remove redundant imports

cmbrennan002 committed:
capitalise global variables and update requirements

cmbrennan002 commented:
Thanks for your comments @RFOxbury! I've addressed them all, and moved the files to a sub-folder in pipeline. 

I had to change the internal tests a fair bit to export a list of errors (but I think it was worthwhile), do you mind looking through it again? Please tell me if I've missed something, your edits looked much cleaner I just couldn't make them work on a pandas df (and thought the cost of moving from running a test on an array to each one individually seemed high).

RFOxbury commented:
> Thanks for your comments @RFOxbury! I've addressed them all, and moved the files to a sub-folder in pipeline.
> 
> I had to change the internal tests a fair bit to export a list of errors (but I think it was worthwhile), do you mind looking through it again? Please tell me if I've missed something, your edits looked much cleaner I just couldn't make them work on a pandas df (and thought the cost of moving from running a test on an array to each one individually seemed high).

Hey @cmbrennan002 ! This looks good - I think it's nice that you now save a file of job IDs where the salary conditions aren't met. Happy for you to merge!

cmbrennan002 committed:
Ensuring notebook outputs are clear prior to merge
